cpp/lib/TextUtil.cc

```diff
@@ -172,11 +176,55 @@ bool ChopIntoWords(const std::string &text, std::unordered_set<std::string> * co
     if (word.length() >= min_word_length) {
         if (unlikely(not WCharToUTF8String(word, &utf8_word)))
             return false;
-        words->insert(utf8_word);
+        words->insert(words->end(), utf8_word);
     }
 
     return true;
 }
+
+
+} // unnamed namespace
+
+
+bool ChopIntoWords(const std::string &text, std::unordered_set<std::string> * const words,
+                   const unsigned min_word_length)
+{
+    return ChopIntoWords<std::unordered_set<std::string>> (text, words, min_word_length);
+}
+
+
+bool ChopIntoWords(const std::string &text, std::vector<std::string> * const words,
+                   const unsigned min_word_length)
+{
+    return ChopIntoWords<std::vector<std::string>> (text, words, min_word_length);
+}
+    
+
+std::vector<std::string>::const_iterator FindSubstring(const std::vector<std::string> &haystack,
+						       const std::vector<std::string> &needle)
+{
+    if (needle.empty())
+	return haystack.cbegin();
+
+    if (haystack.size() < needle.size())
+	return haystack.cend();
+
+    const std::vector<std::string>::const_iterator haystack_start(
+	std::find(haystack.cbegin(), haystack.cend(), needle[0]));
```

Comment:
You have to check for every occurrence of the needle[0] and then for the rest of needle.